.cursorrules

```diff
@@ -0,0 +1,489 @@
+
+2. Complex types (pointer-based):
+
+```cairo
+struct U256 {
+    value: U256Struct*,  // null when pointer is 0
+}
+```
+
+A `null` U256 is represented by the inner U256Struct\* having a value of 0.
+
+### Union Types Pattern
+
+Unions implement a pointer-based variant system. Real example from RLP encoding,
+`Union[Sequence["Simple"], bytes]` and
+`Union[Sequence["Extended"], bytearray, bytes, uint, FixedUnsigned, str, bool]`:
+
+```cairo
+// Simple variant with two possibilities
+struct Simple {
+    value: SimpleEnum*,
+}
+
+struct SimpleEnum {
+    sequence: SequenceSimple,  // First variant
+    bytes: Bytes,             // Second variant
+}
+
+// Extended variant with multiple possibilities
+struct Extended {
+    value: ExtendedEnum*,
+}
+
+struct ExtendedEnum {
+    sequence: SequenceExtended,
+    bytearray: Bytes,
+    bytes: Bytes,
+    uint: Uint*,
+    fixed_uint: Uint*,
+    str: String,
+    bool: Bool*,
+}
+```
+
+Implementation pattern:
+
+1. Outer struct contains pointer to enum struct
+2. Enum struct contains all possible variants.
+3. Only one variant is active (non-zero) at a time
+4. Variant construction uses explicit constructors:
+
+```cairo
+// Example constructor for a variant
+func bytes(value: Bytes) -> Extended {
+    tempvar extended = Extended(
+        new ExtendedEnum(
+            sequence=SequenceExtended(cast(0, SequenceExtendedStruct*)),
+            bytearray=Bytes(cast(0, BytesStruct*)),
+            bytes=value,  // Active variant
+            uint=cast(0, Uint*),
+            fixed_uint=cast(0, Uint*),
+            str=String(cast(0, StringStruct*)),
+            bool=cast(0, Bool*),
+        ),
+    );
+    return extended;
+}
+```
+
+### Write-once Collections Pattern
+
+The inner struct is a struct with two fields: a pointer to the collection's data
+segment, and the length of the collection.
+
+```cairo
+struct Bytes {
+    value: BytesStruct*,
+}
+
+struct BytesStruct {
+    data: felt*,
+    len: felt
+}
+```
+
+We can also make collections of collections, like a tuple of bytes.
+
+```cairo
+struct TupleBytesStruct {
+    data: Bytes*,  // Tuple data
+    len: felt, // Number of elements in the tuple
+}
+
+struct TupleBytes {
+    value: TupleBytesStruct*,
+}
+```
+
+### Mappings Pattern
+
+Implementation using DictAccess pattern from the standard library. For the
+`Mapping[Bytes, Bytes]` type, the corresponding Cairo struct is:
+
+```cairo
+struct BytesBytesDictAccess {
+    key: Bytes,
+    prev_value: Bytes,
+    new_value: Bytes,
+}
+
+struct MappingBytesBytesStruct {
+    dict_ptr_start: BytesBytesDictAccess*,
+    dict_ptr: BytesBytesDictAccess*,
+}
+
+struct MappingBytesBytes {
+    value: MappingBytesBytesStruct*,
+}
+```
+
+Key implementation notes:
+
+- Keys stored as pointers, thus querying a key is a pointer equality check, not
+  a value equality check.
+- In the future, we might consider using key hashes for content-based comparison
+  rather than pointer comparison.
```

Comment:
```suggestion
- Keys are stored as pointers, thus querying a key is a pointer equality check, not
  a value equality check.
- In the future, we consider using key hashes for content-based comparison
  rather than pointer comparison.
```